src/libcollections/btree/node.rs

```diff
@@ -1098,8 +1094,60 @@ impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Internal>, marker::
                 ForceResult::Internal(mut internal) => internal.push(k, v, edge.unwrap())
             }
         }
+    }
+
+    pub fn bulk_steal_left(&mut self, n: usize) {
+        unsafe {
+            // Make room for stolen elements in right child.
+            let mut right = self.reborrow_mut().right_edge().descend();
+            let old_len = right.len();
+
+            debug_assert!(old_len + n <= CAPACITY);
+
+            right.as_leaf_mut().len += n as u16;
+            ptr::copy(right.keys().as_ptr(),
+                      right.keys_mut().as_mut_ptr().offset(n as isize),
+                      old_len);
+            ptr::copy(right.vals().as_ptr(),
+                      right.vals_mut().as_mut_ptr().offset(n as isize),
+                      old_len);
+            if let ForceResult::Internal(mut node) = right.force() {
+                ptr::copy(node.as_internal().edges.as_ptr(),
+                          node.as_internal_mut().edges.as_mut_ptr().offset(n as isize),
+                          old_len+1);
+                for i in 0..old_len+1 {
+                    Handle::new_edge(node.reborrow_mut(), n+i).correct_parent_link();
+                }
+            }
+        }
 
-        self
+        unsafe {
+            // Move elements from the left child to the right one.
+            for i in 0..n {
```

Comment:
I think it'd be faster and more readable to use `ptr::copy_nonoverlapping` for this. Dealing with lifetimes is a bit tricky, as you simultaneously need mutable pointers to the current key/value pair, the left node, and the right node, but that should be possible by just storing them as raw pointers and dereferencing them after you've stored all the pointers.